File: api/Data/Slug.cs
using System.Text;

namespace CbdgApi.Data;

/// <summary>
/// Generates URL-friendly slugs from category names. Unicode letters/digits
/// are kept as-is (not transliterated to ASCII) - CBDG is bilingual (en/he)
/// and a Hebrew-named book deserves a real, readable slug in its own script,
/// not a GUID just because it isn't Latin text. A name that slugifies to
/// nothing (symbols/emoji only) falls back to the category's own id, which
/// keeps the "Slug" column always non-null and always resolvable.
/// </summary>
public static class Slug
{
    public static string From(string name)
    {
        var lower = name.Trim().ToLowerInvariant();
        var sb = new StringBuilder();
        var lastWasHyphen = true; // swallow a would-be leading hyphen
        foreach (var ch in lower)
        {
            if (char.IsLetterOrDigit(ch))
            {
                sb.Append(ch);
                lastWasHyphen = false;
            }
            else if (!lastWasHyphen)
            {
                sb.Append('-');
                lastWasHyphen = true;
            }
        }
        return sb.ToString().Trim('-');
    }

    /// <summary>Slugifies `name`, then appends -2, -3, ... until `existsAsync`
    /// reports the candidate is free. `fallbackId` covers the empty-slug case.</summary>
    public static async Task<string> GenerateUniqueAsync(
        string name,
        Guid fallbackId,
        Func<string, Task<bool>> existsAsync)
    {
        var baseSlug = From(name);
        if (baseSlug.Length == 0) return fallbackId.ToString();

        var candidate = baseSlug;
        var suffix = 2;
        while (await existsAsync(candidate))
        {
            candidate = $"{baseSlug}-{suffix}";
            suffix++;
        }
        return candidate;
    }
}

File: src/slug.ts
/**
 * Mirrors api/Data/Slug.cs (kept in sync by hand - no shared package between
 * the two runtimes). Unicode letters/digits are kept as-is rather than
 * transliterated to ASCII, since CBDG is bilingual and a Hebrew book name
 * deserves a real slug in its own script, not a forced-empty one.
 */
export function slugify(name: string): string {
  const lower = name.trim().toLowerCase()
  let result = ''
  let lastWasHyphen = true // swallow a would-be leading hyphen
  for (const ch of lower) {
    if (/[\p{L}\p{N}]/u.test(ch)) {
      result += ch
      lastWasHyphen = false
    } else if (!lastWasHyphen) {
      result += '-'
      lastWasHyphen = true
    }
  }
  return result.replace(/^-+|-+$/g, '')
}

/** Slugifies `name`, then appends -2, -3, ... until `exists` reports the
 * candidate is free. `fallbackId` covers the empty-slug case (symbols/emoji-only names). */
export function generateUniqueSlug(
  name: string,
  fallbackId: string,
  exists: (candidate: string) => boolean,
): string {
  const base = slugify(name)
  if (!base) return fallbackId

  let candidate = base
  let suffix = 2
  while (exists(candidate)) {
    candidate = `${base}-${suffix}`
    suffix++
  }
  return candidate
}
